firedrake/slate/slate.py

```diff
@@ -0,0 +1,767 @@
+    def _compute_renumbering(self):
+        """Renumbers coefficients by including integration domains."""
+
+        dn = self._domain_numering
+        cn = self._coefficient_numbering
+        renumbering = {}
+        renumbering.update(dn)
+        renumbering.update(cn)
+
+        n = len(dn)
+        for c in cn:
+            d = c.ufl_domain()
+            if d is not None and d not in renumbering:
+                renumbering[d] = n
+                n += 1
+
+        return renumbering
+
+    def _compute_signature(self, renumbering):
+        """Computes the signature of the tensor from the integrals."""
+        # TODO: Is there a better/cheaper way of doing this?!
+
+        integrals = self._integrals
+        integrands = [integral.integrand() for integral in integrals]
+
+        terminal_hashdata = compute_terminal_hashdata(integrands, renumbering)
+
+        hashdata = []
+        for integral in integrals:
+            integrand_hashdata = compute_expression_hashdata(integral.integrand(),
+                                                             terminal_hashdata)
+            domain_hashdata = integral.ufl_domain()._ufl_signature_data_(renumbering)
+
+            integral_hashdata = (integrand_hashdata, domain_hashdata,
+                                 integral.integral_type(),
+                                 integral.subdomain_id(),
+                                 canonicalize_metadata(integral.metadata()), )
+            hashdata.append(integral_hashdata)
+
+        return hashlib.sha512(str(hashdata).encode('utf-8')).hexdigest()
+
+    def signature(self):
+        """Signature for use in caching. Reproduced from ufl."""
+
+        if self._signature is None:
+            self._signature = self._compute_signature(self._compute_renumbering())
+
+        return self._signature
+
+    def __hash__(self):
+        """Returns a hash code for use in dictionary objects."""
+        if self._hash is None:
+            self._hash = hash((type(self), )
+                              + tuple(hash(it) for it in self._integrals))
+
+        return self._hash
+
+
+class Scalar(Tensor):
+    """An abstract representation of a rank-0 tensor object in SLATE."""
+
+    def __init__(self, form):
+        """Constructor for the Scalar class.
+
+        : arg form: a ufl form representing a scalar object:
+                    i.e. form = Constant(1.0, domain=mesh)*dx.
+        """
+
+        r = len(form.arguments())
+        if r != 0:
+            rank_error(0, r)
+        self.form = form
+        self.check_integrals(form.integrals())
+        Tensor.tensor_id += 1
+        super(Scalar, self).__init__(arguments=(),
+                                     coefficients=form.coefficients(),
+                                     integrals=form.integrals())
+
+    def __str__(self, prec=None):
+        """String representation of a SLATE Scalar object."""
+        return "S_%d" % self.tensor_id
+
+    __repr__ = __str__
```

Comment:
Normally `__repr__` is something that could be used to reconstruct the object.  So perhaps `"Scalar(%r)" % self.form`